src/KafkaFlow/Consumers/PartitionOffsets.cs

```diff
@@ -42,16 +42,24 @@ public bool TryDequeue(IConsumerContext context)
                 {
                     this.DequeuedContext = _receivedContexts.First.Value;
                     _receivedContexts.RemoveFirst();
-                }
-                while (_receivedContexts.Count > 0 && _processedContexts.Remove(_receivedContexts.First.Value.Offset));
+                } while (_receivedContexts.Count > 0 && _processedContexts.Remove(_receivedContexts.First.Value.Offset));
             }
 
             return true;
         }
 
-        public Task WaitContextsCompletionAsync() => Task.WhenAll(
-            _receivedContexts
-                .Select(x => x.Completion)
-                .ToList());
+        public Task WaitContextsCompletionAsync()
+        {
+            List<Task> tasks;
+
+            lock (_receivedContexts)
+            {
+                tasks = _receivedContexts
+                    .Select(x => (Task)x.Completion)
+                    .ToList();
```

Comment:
The to List is used to materialize the Tasks and avoid an exception if the LinkedList is changed during its iteration.

Reply:
Why not to use an Queue<T> here

also you can use something like

```csharp
  lock (_receivedContexts)
  {
return Task.WhenAll(_receivedContexts
                    .Select(x => (Task)x.Completion))
}
```

Reply:
A LinkedList can work as a FIFO without the need to expand.

Regarding the use of WhenAll inside the lock, it is potentially risky as we are uncertain about the inner workings of the WhenAll method and any changes to its behavior could have unintended consequences. In the worst-case scenario, we may end up releasing the lock before the Task enumerator has been fully iterated.

Reply:
> Why not to use an Queue here?
> 
> also you can use something like
> 
> ```cs
> lock (_receivedContexts)
> {
>   return Task.WhenAll(_receivedContexts.Select(x => (Task)x.Completion))
> }
> ```

This would be fine if inside the lock the Task.WhenAll would be awaited (with the cost of creating the async/await state machine). Without the await, @filipeesch is on point.